File: packages/ts/tests/e2e/fanout.test.ts
// CLI e2e: fan-out file->table matching through the real Node launcher +
// binary. A file matching two tables' overlapping globs populates both tables;
// querying each over the real `POST /query` surface returns the file's row
// (#580). No mocks: real launcher, real binary, real process, real filesystem.
//
// The binary is staged under a PRIVATE temp `node_modules` handed to the
// launcher via NODE_PATH (honored by the launcher's CJS `require.resolve`),
// rather than the package's own `node_modules` (which other e2e files
// stage/remove and vitest runs in parallel).

import { type ChildProcess, execFileSync, spawn } from "node:child_process";
import { mkdir, mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import http from "node:http";
import net from "node:net";
import { tmpdir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { afterAll, beforeAll, describe, expect, it } from "vitest";
import { PLATFORMS, librarySlug } from "../../src/platforms.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PKG_ROOT = join(__dirname, "..", "..");

const HOST = PLATFORMS.find(
  (p) =>
    `${p.nodePlatform}-${p.nodeArch}` === `${process.platform}-${process.arch}`,
);
if (!HOST) {
  throw new Error(`unsupported host ${process.platform}-${process.arch}`);
}
const BINARY = join(
  PKG_ROOT,
  "build",
  `bundled-cli-${librarySlug(HOST)}`,
  HOST.exe ? "dirsql.exe" : "dirsql",
);

let LAUNCHER: string;
let STAGE_ROOT: string;
let NODE_PATH_DIR: string;
let serverProc: ChildProcess;
let serverPort: number;
let serverLogs = "";

interface RawResponse {
  status: number;
  text: string;
}

function httpRaw(
  port: number,
  method: "GET" | "POST",
  body?: string,
): Promise<RawResponse> {
  return new Promise((resolve, reject) => {
    const req = http.request(
      {
        host: "127.0.0.1",
        port,
        path: "/query",
        method,
        headers:
          body === undefined
            ? {}
            : {
                "content-type": "application/json",
                "content-length": Buffer.byteLength(body),
              },
      },
      (res) => {
        let data = "";
        res.on("data", (chunk) => {
          data += chunk;
        });
        res.on("end", () =>
          resolve({ status: res.statusCode ?? 0, text: data }),
        );
      },
    );
    req.on("error", reject);
    if (body !== undefined) {
      req.write(body);
    }
    req.end();
  });
}

async function postSql(port: number, sql: string): Promise<RawResponse> {
  return httpRaw(port, "POST", JSON.stringify({ sql }));
}

function freePort(): Promise<number> {
  return new Promise((resolve) => {
    const srv = net.createServer();
    srv.listen(0, "127.0.0.1", () => {
      const { port } = srv.address() as net.AddressInfo;
      srv.close(() => resolve(port));
    });
  });
}

function waitForServer(proc: ChildProcess, port: number, timeout = 5_000) {
  return new Promise<boolean>((resolve) => {
    let done = false;
    const fin = (v: boolean) => {
      if (!done) {
        done = true;
        resolve(v);
      }
    };
    setTimeout(() => fin(false), timeout);
    proc.once("exit", () => fin(false));
    (function attempt() {
      if (done) {
        return;
      }
      const sock = net.createConnection({ host: "127.0.0.1", port });
      sock.on("connect", () => {
        sock.destroy();
        fin(true);
      });
      sock.on("error", () => {
        sock.destroy();
        if (!done) {
          setTimeout(attempt, 50);
        }
      });
    })();
  });
}

beforeAll(async () => {
  const pkg = JSON.parse(
    await readFile(join(PKG_ROOT, "package.json"), "utf8"),
  ) as { bin: { dirsql: string } };
  LAUNCHER = join(PKG_ROOT, pkg.bin.dirsql);

  STAGE_ROOT = await mkdtemp(join(tmpdir(), "dirsql-fanout-e2e-"));
  NODE_PATH_DIR = join(STAGE_ROOT, "node_modules");
  const cliPkgDir = join(NODE_PATH_DIR, HOST.name);
  const binName = HOST.exe ? "dirsql.exe" : "dirsql";
  await mkdir(cliPkgDir, { recursive: true });
  await writeFile(
    join(cliPkgDir, "package.json"),
    JSON.stringify({ name: HOST.name, version: "0.0.0" }),
  );
  execFileSync("cp", [BINARY, join(cliPkgDir, binName)]);

  const dataDir = join(STAGE_ROOT, "data");
  await mkdir(join(dataDir, "data", "2401.00001"), { recursive: true });
  await writeFile(join(dataDir, "data", "2401.00001", "metadata.json"), "{}");
  const cfg = join(dataDir, ".dirsql.toml");
  // Each table emits its own `path` column via an `on-file` hook (the core no
  // longer injects filesystem facts): strip the root prefix and print it.
  const pathHook = `on-file = '''sh -c 'rel=\${1#"$2"/}; printf "[{\\"path\\":\\"%s\\"}]" "$rel"' sh {path} {root}'''`;
  await writeFile(
    cfg,
    `[[table]]\nddl = "CREATE TABLE ta (path TEXT)"\nglob = "data/*/metadata.json"\n${pathHook}\n\n[[table]]\nddl = "CREATE TABLE tb (path TEXT)"\nglob = "data/**/metadata.json"\n${pathHook}\n`,
  );

  serverPort = await freePort();
  serverProc = spawn(
    process.execPath,
    [
      LAUNCHER,
      "server",
      "--config",
      cfg,
      "--host",
      "127.0.0.1",
      "--port",
      String(serverPort),
    ],
    {
      stdio: "pipe",
      cwd: dataDir,
      env: {
        ...process.env,
        NODE_PATH: [NODE_PATH_DIR, process.env.NODE_PATH]
          .filter(Boolean)
          .join(":"),
      },
    },
  );
  serverProc.stdout?.on("data", (c: Buffer) => {
    serverLogs += c.toString();
  });
  serverProc.stderr?.on("data", (c: Buffer) => {
    serverLogs += c.toString();
  });
  const ok = await waitForServer(serverProc, serverPort);
  if (!ok) {
    throw new Error(`dirsql server did not start\n${serverLogs}`);
  }
}, 60_000);

afterAll(async () => {
  serverProc?.kill("SIGTERM");
  if (serverProc) {
    await new Promise<void>((resolve) =>
      serverProc.once("exit", () => resolve()),
    );
  }
  await rm(STAGE_ROOT, { recursive: true, force: true });
});

describe("fan-out file->table matching (#580)", () => {
  it("populates the first-declared table", async () => {
    const res = await postSql(serverPort, "SELECT path FROM ta");
    expect(res.status).toBe(200);
    const rows = JSON.parse(res.text) as { path: string }[];
    expect(rows.map((r) => r.path)).toEqual(["data/2401.00001/metadata.json"]);
  });

  it("also populates the second-declared table", async () => {
    const res = await postSql(serverPort, "SELECT path FROM tb");
    expect(res.status).toBe(200);
    const rows = JSON.parse(res.text) as { path: string }[];
    expect(rows.map((r) => r.path)).toEqual(["data/2401.00001/metadata.json"]);
  });
});

File: packages/python/tests/e2e/fanout_test.py
"""CLI e2e: fan-out file->table matching through the real launcher + binary.

A file matching two tables' overlapping globs populates both tables; querying
each over the real `POST /query` surface returns the file's row (#580). No
mocks: real launcher, real binary, real process, real filesystem.
"""

import json
import os
import shutil
import socket
import subprocess
import sys
import time
import urllib.error
import urllib.request

import pytest

_REPO_ROOT = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..", "..", "..")
)
_RELEASE = os.path.join(_REPO_ROOT, "target", "release", "dirsql")
_DEBUG = os.path.join(_REPO_ROOT, "target", "debug", "dirsql")
_BINARY = _RELEASE if os.path.exists(_RELEASE) else _DEBUG

import dirsql as _dirsql_pkg

_BINARY_STAGE_DIR = os.path.join(os.path.dirname(_dirsql_pkg.__file__), "_binary")

# An `on-file` hook emitting `path`, derived from the file path (`{path}`)
# relative to the scan root (`{root}`).
_HOOK_PATH = r"""on-file = '''sh -c 'rel=${1#"$2"/}; printf "[{\"path\":\"%s\"}]" "$rel"' sh {path} {root}'''"""


def _free_port():
    with socket.socket() as s:
        s.bind(("", 0))
        return s.getsockname()[1]


def _wait_for_server(proc, port, timeout=5.0):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        if proc.poll() is not None:
            return False
        try:
            urllib.request.urlopen(f"http://localhost:{port}/query", timeout=0.5)
            return True
        except urllib.error.HTTPError:
            return True
        except Exception:
            time.sleep(0.05)
    return False


def _post_sql(port, sql):
    req = urllib.request.Request(
        f"http://localhost:{port}/query",
        data=json.dumps({"sql": sql}).encode(),
        headers={"content-type": "application/json"},
    )
    try:
        with urllib.request.urlopen(req) as resp:
            return resp.status, json.loads(resp.read())
    except urllib.error.HTTPError as err:
        return err.code, json.loads(err.read())


def describe_fanout():
    @pytest.fixture
    def server(tmp_path):
        assert os.path.exists(_BINARY), (
            f"dirsql binary not built at {_BINARY}; "
            "run `cargo build -p dirsql --features cli` first"
        )
        os.makedirs(_BINARY_STAGE_DIR, exist_ok=True)
        staged_binary = os.path.join(_BINARY_STAGE_DIR, "dirsql")
        shutil.copy(_BINARY, staged_binary)
        os.chmod(staged_binary, 0o755)

        root = tmp_path / "data"
        sub = root / "data" / "2401.00001"
        sub.mkdir(parents=True)
        (sub / "metadata.json").write_text("{}")
        cfg = root / ".dirsql.toml"
        cfg.write_text(
            "[[table]]\n"
            'name = "ta"\n'
            'ddl = "CREATE TABLE ta (path TEXT)"\n'
            'glob = "data/*/metadata.json"\n'
            f"{_HOOK_PATH}\n"
            "\n"
            "[[table]]\n"
            'name = "tb"\n'
            'ddl = "CREATE TABLE tb (path TEXT)"\n'
            'glob = "data/**/metadata.json"\n'
            f"{_HOOK_PATH}\n"
        )

        port = _free_port()
        proc = subprocess.Popen(
            [
                sys.executable,
                "-c",
                "import sys; from dirsql.cli.main import main; sys.exit(main())",
                "server",
                "--config",
                str(cfg),
                "--host",
                "127.0.0.1",
                "--port",
                str(port),
            ],
            cwd=str(root),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        try:
            ok = _wait_for_server(proc, port)
            if not ok:
                out, err = proc.communicate(timeout=5)
                raise AssertionError(
                    f"server did not start\n--- stdout ---\n{out}\n--- stderr ---\n{err}"
                )
            yield port
        finally:
            proc.terminate()
            proc.wait(timeout=5)
            shutil.rmtree(_BINARY_STAGE_DIR, ignore_errors=True)

    def it_populates_the_first_declared_table(server):
        status, rows = _post_sql(server, "SELECT path FROM ta")
        assert status == 200
        assert [r["path"] for r in rows] == ["data/2401.00001/metadata.json"]

    def it_also_populates_the_second_declared_table(server):
        status, rows = _post_sql(server, "SELECT path FROM tb")
        assert status == 200
        assert [r["path"] for r in rows] == ["data/2401.00001/metadata.json"]
